x-pack/plugins/discover_log_explorer/common/immutable_cache.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+export interface IImmutableCache<KeyType, ValueType> {
+  get(key: KeyType): ValueType | undefined;
+  set(key: KeyType, value: ValueType): ImmutableCache<KeyType, ValueType>;
+  has(key: KeyType): boolean;
+  clear(): ImmutableCache<KeyType, ValueType>;
+}
+
+type Cache<ValueType> = Record<string, ValueType>;
+
+export class ImmutableCache<KeyType, ValueType> implements IImmutableCache<KeyType, ValueType> {
+  private cache: Cache<ValueType>;
+
+  constructor(cache?: Cache<ValueType>) {
+    this.cache = cache ?? {};
+  }
+
+  public get(key: KeyType): ValueType | undefined {
+    const serializedKey = this.serialize(key);
+    return this.cache[serializedKey];
+  }
+
+  public set(key: KeyType, value: ValueType): ImmutableCache<KeyType, ValueType> {
+    const serializedKey = this.serialize(key);
+    this.cache[serializedKey] = value;
+    return new ImmutableCache(this.cache);
```

Comment:
Doesn't this mean that the old cache instance is mutated in place because they share internal state?